[Do not merge] Prototype: Async rewriter alternative 1 (Rust, WASM, no AST manipulation)

addaleax opened the pull request.

addaleax committed:
WIP

addaleax committed:
fixup: VecDeque

addaleax committed:
fixup: IIFE wrapping WIP

addaleax committed:
fixup: ...

kmruiz commented on packages/async-rewriter3/src/lib.rs:
```diff
@@ -0,0 +1,400 @@
+}
+
+fn is_block(body: &ExprOrBlock) -> bool {
+    match body {
```
If you want to make it slimmer (just for fun) you can use matches!

It would be like (with typos for sure):


```rs
matches!(body, ExprOrBlock::Block(_))
```

kmruiz commented on packages/async-rewriter3/src/lib.rs:
```diff
@@ -0,0 +1,400 @@
+                    ret.append(&mut add_all_variables_from_declaration([&pat.unwrap()].into_iter()));
+                }
+            },
+            Pattern::AssignPattern(p) => {
+                let key = p.key();
+                if key.is_some() {
+                    ret.append(&mut add_all_variables_from_declaration([&key.unwrap()].into_iter()));
+                }
+            },
```
You can also use match here to make it simpler to read and avoid that unwrap:

```rs
match p.key()  {
   Some(key) => ret.append(...key)
   _ => 
```
Or even an if with pattern extraction:

```rs
if let Some(key) = p.key {
  ret.append...
}
```

addaleax committed:
fixup: Kevin’s suggestions

addaleax committed:
fixup: invert condition so the difference shows in CI perf tests

addaleax committed:
fixup: include compiled WASM in compile_ts

addaleax committed:
fixup: Rust in CI ...

addaleax committed:
fixup: regenerate automatically created files after rebase

addaleax committed:
fixup: handle e.g. `function foo(a=1){}`

addaleax committed:
WIP

addaleax committed:
fixup: snapshot fix

addaleax committed:
fixup: literal object keys work

addaleax committed:
fixup: template elements should not receive tags

addaleax committed:
fixup: account for typeof + parentheses

addaleax committed:
fixup: set sync rv for arrow functions

addaleax committed:
fixup: handle for-of/in lvalues, labels

addaleax committed:
fixup: handle bracketed call expressions

addaleax committed:
fixup: slash merging

addaleax committed:
fixup: perf

addaleax committed:
fixup: unnecessary .iter()

addaleax committed:
fixup: initialize immediately after class def

addaleax committed:
fixup: extend ; safeguarding

addaleax committed:
fixup: more semicolon safeguards

addaleax committed:
fixup: Rust cleanups

addaleax committed:
fixup: explicitly install wasm-pack

addaleax committed:
fixup: add rust executable variant

addaleax committed:
fixup: switch to oxc parser

addaleax committed:
fixup: post-rewrite fixes

addaleax committed:
fixup: `cargo fmt`

addaleax committed:
fixup: readability

addaleax committed:
fixup: fmt

addaleax committed:
fixup: add c addon variant

addaleax committed:
fixup: typo discovered during walkthrough :)

addaleax committed:
Merge remote-tracking branch 'origin/main' into async-rewriter3

addaleax committed:
Merge remote-tracking branch 'origin/main' into async-rewriter3

addaleax committed:
fixup: account for void patterns in async-rewriter2

addaleax committed:
fixup: bump Cargo.lock

addaleax committed:
fixup: bump deps in Cargo.toml

addaleax committed:
Merge remote-tracking branch 'origin/main' into async-rewriter3

addaleax committed:
fixup: update dependencies of async-rewriter3 to latest

addaleax committed:
chore: split out async-rewriter tests into separate package

This will be a prerequisite for productionizing a new iteration
of the async rewriter, such as the one proposed in
https://github.com/mongodb-js/mongosh/pull/1931.

addaleax committed:
fixup: copilot CR

addaleax committed:
fixup: knip

addaleax committed:
Merge branch 'async-rewriter-split-out-tests' into async-rewriter3

addaleax committed:
Merge remote-tracking branch 'origin/main' into async-rewriter3

addaleax committed:
fixup: LLM prototype for iteration

addaleax committed:
fixup: remove custom strict mode checking

addaleax committed:
fixup: simplify signature

addaleax committed:
fixup: cargo fmt

addaleax committed:
fixup: deduplicate JS source, add back identifier templates

addaleax committed:
fixup: simplify for loop handling

addaleax committed:
fixup: revert wasm-pack profile change, adjust README

addaleax committed:
fixup: remove explicit copy of runtime support code

addaleax committed:
fixup: clean up string encoding functions

addaleax committed:
fixup: function kind handling

addaleax committed:
fixup: simplify try-catch handling

addaleax committed:
fixup: declare dependency ar3 -> ar2 for runtime support script
